spec/controllers/exercises_controller_spec.rb

```diff
@@ -1,92 +1,212 @@
+        allow(controller).to receive(:current_user).and_return(current_user)
+      end
+
+      it 'responds successfully to a GET request' do
+        get :show, params: { id: exercise.id }
+        expect(response).to be_successful
+      end
 
-    it 'assigns the return value of exercise#rehearsals_for_user to the rehearsals ivar' do
-      exercise = FactoryBot.create(:exercise)
-      allow(Exercise).to receive(:find).with(exercise.id.to_s).and_return(exercise)
+      it 'returns a 200 HTTP response code' do
+        get :show, params: { id: exercise.id }
+        expect(response).to have_http_status('200')
+      end
 
-      rehearsals_for_user = double('rehearsals_for_user')
-      allow(exercise).to receive(:rehearsals_for_user).with(user).and_return(rehearsals_for_user)
+      it 'raises an active record error if the exercise does not exist' do
+        expect { get :show, params: { id: '2384' } }.to raise_error(ActiveRecord::RecordNotFound)
+      end
 
-      get :show, id: exercise.id
-      rehearsals_ivar = assigns(:rehearsals)
+      it 'blocks you (redirects you to root) from viewing a private, non-published/non-global exercise that does not belong to you' do
+        not_my_exercise = FactoryBot.create(:exercise, user_id: 35, global: false)
+        get :show, params: { id: not_my_exercise.id }
+        expect(response).to redirect_to root_path
+      end
+    end
+
+    context 'when not logged-in' do
+      it 'returns a 302 response' do
+        get :show, params: { id: exercise.id }
+        expect(response).to have_http_status('302')
+      end
 
-      expect(rehearsals_ivar).to eq(rehearsals_for_user)
+      it 'redirects to the sign-in page' do
+        get :show, params: { id: exercise.id }
+        expect(response).to redirect_to('/users/sign_in')
+      end
     end
+  end
 
-    context 'when the exercise has no tags' do
-      it 'doctrines ivar is nil' do
-        exercise = FactoryBot.create(:exercise)
+  describe '#edit' do
+    context 'As a signed-in, non-admin user' do
+      let(:current_user) { FactoryBot.create(:user) }
+      let(:exercise) { FactoryBot.create(:exercise, user_id: current_user.id) }
 
-        get :show, id: exercise.id
-        doctrines_ivar = assigns(:doctrines)
+      before do
+        allow(controller).to receive(:authenticate_user!)
+        allow(controller).to receive(:current_user).and_return(current_user)
+      end
 
-        expect(doctrines_ivar).to be_nil
+      it 'blocks you (redirects you to root) from editing an exercise that is PRIVATE and does not belong to you' do
+        second_exercise = FactoryBot.create(:exercise, user_id: 35, global: false)
+        get :edit, params: { id: second_exercise.id }
+        expect(response).to redirect_to root_path
+      end
+
+      it 'blocks you (redirects you to root) from editing an exercise that is GLOBAL and does not belong to you' do
+        second_exercise = FactoryBot.create(:exercise, user_id: 35, global: true)
```

Comment:
Assigning IDs like that can become quite a headache (there might suddenly be a `current_user` with ID 35, which will break the spec)

Normally the factory would create a minimal, but complete valid object, so this should not be necessary. If you want to make it explicit, you could do (assuming the shorter syntax mentioned above):

```ruby
second_exercise = create(:exercise, user: create(:user), global: true)
```

Reply:
Yes great point.

OK fixed. I implemented your suggestion in all of these examples. I generate a new user factory for each one instead of manually assigning the user ID.